### ISSUE TITLE
Test Interrupt/inttim/inttim1 fails

### ISSUE THREAD

**Issue by dirkwhoffmann** [dirkwhoffmann/vAmiga#407]
vAmiga has an issue with interrupt timing if INTENA or INTREQ are written by the CPU 🤒.

A500 8A: 🥰
![inttim1_A500_8A](https://user-images.githubusercontent.com/12561945/92485685-8d647100-f1eb-11ea-9d75-27e0a47abb8d.JPG)

vAmiga: 🙈

<img width="900" alt="Bildschirmfoto 2020-09-08 um 15 52 13" src="https://user-images.githubusercontent.com/12561945/92485718-96554280-f1eb-11ea-8f11-d78e018be1bf.png">

**Comment by dirkwhoffmann** [comment]
I've "upgraded" the test a bit (and added five new ones, all failing). The fourth (yellow) color bar in the original image above is interesting. It is produced by the following code

	lea     $DFF09E,a3
	move.w  #$8004,-(a3)    ; Writes into INTREQ
	moveq   #1,d0 
	moveq   #2,d0 
	moveq   #3,d0 

The value of d0 inside the interrupt handler defines the bar color (0 = green, 1 = red, 2 = yellow, 3 = blue). In predecrement  addressing mode (which I used for the move.w), the prefetch cycle happens before the write cycle (for whatever reason). As a result, the IPL line change is not recognized by the CPU before the next two moveq statements are executed. 

vAmiga screws things up badly. In the case mentioned above, it executes three moveq's until it jumps into the IRQ handler. 

Bottom line: My IPL delay pipe seems to work only if the Copper writes to INTREQ or INTENA (all my older test cases utilize the Copper for triggering interrupts). If the CPU writes into these registers, timing is screwed up.

**Comment by mithrendal** [comment]
Great catch 🧑🏻‍🚀

**Comment by dirkwhoffmann** [comment]
> Great catch 🧑🏻‍🚀

If only I knew how it could be fixed 🤓.

**Comment by mithrendal** [comment]
The real amiga goes into interrupt a bit earlier... No?  Or the other way around Moira keeps on executing to fast ... is being interrupted to late ... no?

What means IPL, I already forgot that? Interrupt Pipeline?

I bet you need to meditate like a guru 👳🏽‍♂️

**Comment by dirkwhoffmann** [comment]
IPL = Interrupt Priority Line = Requested interrupt level by Paula (1 = lowest, 7 = highest (NMI), Amigas don't use level 7 though)

>  Or the other way around Moira keeps on executing to fast 

No, when INTREQ or INTENA is changed, it takes some time until the CPU recognizes the interrupt. All delays seem to be fine for Copper writes, but the delay is too large for CPU writes. I am doing experiments with some timing combinations right now...

**Comment by emoon** [comment]
Did you test this in UAE as well and see if it behaves correct there?

**Comment by dirkwhoffmann** [comment]
All test cases pass when the IPL pipe delay is reduced to three and Copper writes get one additional delay cycle. Hence, this code change does the trick: 

```
template <Accessor s> void
Paula::pokeINTENA(u16 value)
{
    debug(INTREG_DEBUG, "pokeINTENA(%X)\n", value);

    if (s == CPU_ACCESS) {
        paula.setINTENA(value);
    } else {
        agnus.recordRegisterChange(DMA_CYCLES(1), SET_INTENA, value);
    }
}
```

The IPL pipe delay is handled in setINTENA(), resp. setINTREQ().

I've released v0.9.11 today with this fix, because interrupt timing is a crucial aspect. Such changes have the habit to break other stuff.

**Comment by emoon** [comment]
Ah, Does this solve #348 issue?

**Comment by dirkwhoffmann** [comment]
> Did you test this in UAE as well and see if it behaves correct there?

Yes, UAE passes inttim1 to inttim6 without any issues. Now, vAmiga 0.9.11 passes these tests, too (without breaking any of the existing tests). 

> Ah, Does this solve #348 issue?

No, unfortunately not. The Jetsons are tougher than the Adams family. 🙁

**Comment by emoon** [comment]
Ah, too bad
